libcudacxx/include/cuda/std/detail/libcxx/include/variant

```diff
@@ -333,294 +438,331 @@ constexpr size_t __find_index() {
+    {__move_constructible_trait, __trait<_Types, is_trivially_move_assignable, is_move_assignable>...});
 
-  static constexpr _Trait __destructible_trait = __common_trait(
-      {__trait<_Types, is_trivially_destructible, is_destructible>...});
+  static constexpr _Trait __destructible_trait =
+    __variant_detail::__common_trait({__trait<_Types, is_trivially_destructible, is_destructible>...});
 };
 
-namespace __access {
+namespace __access
+{
 
-struct __union {
+struct __union
+{
   template <class _Vp>
-  inline _LIBCUDACXX_INLINE_VISIBILITY
-  static constexpr auto&& __get_alt(_Vp&& __v, in_place_index_t<0>) {
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr auto&& __get_alt(_Vp&& __v, in_place_index_t<0>)
+  {
     return _CUDA_VSTD::forward<_Vp>(__v).__head;
   }
 
   template <class _Vp, size_t _Ip>
-  inline _LIBCUDACXX_INLINE_VISIBILITY
-  static constexpr auto&& __get_alt(_Vp&& __v, in_place_index_t<_Ip>) {
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr auto&& __get_alt(_Vp&& __v, in_place_index_t<_Ip>)
+  {
     return __get_alt(_CUDA_VSTD::forward<_Vp>(__v).__tail, in_place_index<_Ip - 1>);
   }
 };
 
-struct __base {
+struct __base
+{
   template <size_t _Ip, class _Vp>
-  inline _LIBCUDACXX_INLINE_VISIBILITY
-  static constexpr auto&& __get_alt(_Vp&& __v) {
-    return __union::__get_alt(_CUDA_VSTD::forward<_Vp>(__v).__data,
-                              in_place_index<_Ip>);
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr auto&& __get_alt(_Vp&& __v)
+  {
+    return __union::__get_alt(_CUDA_VSTD::forward<_Vp>(__v).__data, in_place_index<_Ip>);
   }
 };
 
-struct __variant {
+struct __variant
+{
   template <size_t _Ip, class _Vp>
-  inline _LIBCUDACXX_INLINE_VISIBILITY
-  static constexpr auto&& __get_alt(_Vp&& __v) {
-    return __base::__get_alt<_Ip>(_CUDA_VSTD::forward<_Vp>(__v).__impl);
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr auto&& __get_alt(_Vp&& __v)
+  {
+    return __base::__get_alt<_Ip>(_CUDA_VSTD::forward<_Vp>(__v).__impl_);
   }
 };
 
 } // namespace __access
 
-namespace __visitation {
+namespace __visitation
+{
 
-struct __base {
+struct __base
+{
   template <class _Visitor, class... _Vs>
-  inline _LIBCUDACXX_INLINE_VISIBILITY
-  static constexpr decltype(auto)
-  __visit_alt_at(size_t __index, _Visitor&& __visitor, _Vs&&... __vs) {
+  _LIBCUDACXX_INLINE_VISIBILITY static constexpr decltype(auto)
+  __visit_alt_at(size_t __index, _Visitor&& __visitor, _Vs&&... __vs)
+  {
     constexpr auto __fdiagonal =
-        __make_fdiagonal<_Visitor&&,
-                         decltype(_CUDA_VSTD::forward<_Vs>(__vs).__as_base())...>();
-    return __fdiagonal[__index](_CUDA_VSTD::forward<_Visitor>(__visitor),
-                                _CUDA_VSTD::forward<_Vs>(__vs).__as_base()...);
+      __make_fdiagonal<_Visitor&&, decltype(_CUDA_VSTD::forward<_Vs>(__vs).__as_base())...>();
+    return __fdiagonal[__index](_CUDA_VSTD::forward<_Visitor>(__visitor), _CUDA_VSTD::forward<_Vs>(__vs).__as_base()...);
```

Comment:
important: this approach does not perform well. Index is a runtime variable here. When array is indexed at runtime, array gets spilled into local memory from registers. This creates a lot of memory traffic. Sample benchmark below compares performance of recursive visit vs stack array approach used in this PR:
```cpp
template <class... Ts>
__device__ void visit(cuda::std::variant<Ts...> &v, int *result) {
  cuda::std::visit([result](auto arg) {
    if (static_cast<int>(arg) > 42) {
      atomicAdd(result, 1);
    }
  }, v);
}

template <class... Ts>
__device__ void visit(stdexec::variant_t<Ts...> &v, int *result) {
  stdexec::visit([result](auto arg) {
    if (static_cast<int>(arg) > 42) {
      atomicAdd(result, 1);
    }
  }, v);
}

template <class VariantT>
__global__ void kernel(VariantT *v, int *result) {
  visit(*v, result);
}

template <class VariantT>
float test() {
  int *err{};
  VariantT *v{};
  cudaMallocManaged(&v, sizeof(VariantT));
  cudaMallocManaged(&err, sizeof(int));
  new (v) VariantT{};
  v->template emplace<char>(42);
  *err = 0;

  cudaEvent_t start, end;
  cudaEventCreate(&start);
  cudaEventCreate(&end);

  cudaEventRecord(start);
  kernel<<<128 * 1024 * 1024, 256>>>(v, err);
  cudaEventRecord(end);
  cudaEventSynchronize(end);

  float ms{};
  cudaEventElapsedTime(&ms, start, end);

  cudaFree(err);
  cudaFree(v);

  cudaEventDestroy(start);
  cudaEventDestroy(end);

  return ms;
}

template <class... Ts>
void speedup()
{
  std::cout << sizeof...(Ts) << ", " << test<cuda::std::variant<Ts...>>() / test<stdexec::variant_t<Ts...>>() << std::endl;
}

int main() {
  speedup<char, unsigned char>();
  speedup<char, unsigned char, short>();
  speedup<char, unsigned char, short, unsigned short>();
  speedup<char, unsigned char, short, unsigned short, int>();
  speedup<char, unsigned char, short, unsigned short, int, unsigned int>();
  speedup<char, unsigned char, short, unsigned short, int, unsigned int, float>();
  speedup<char, unsigned char, short, unsigned short, int, unsigned int, float, double>();
} 
```
On A6000 recursive visit leads to the following speedup:
![image](https://github.com/NVIDIA/cccl/assets/9890394/789865f4-6381-41c6-933d-08d97a176e54)

Reply:
That is an interesting take. I will ponder it, especially given that it has the potential to unlock MSVC 2017 because that breaks with the __fdiagonal approach